### PR TITLE
Adding ME/PS merging to Pythia8Interface 

### PR DESCRIPTION
**Pull request by selvaggi**

### CODE REVIEW

**Comment by jlingema** [comment]
ok to test

File: Generation/src/components/PythiaInterface.cpp [hunk 1]
```diff
@@ -80,13 +137,88 @@ StatusCode PythiaInterface::execute() {
...
+    // we may tag this event accordingly when histogramming. Note
+    // that for MLM jet matching, this might not coincide with the
+    // actual number of partons in the input LH event, since some
+    // partons may be excluded from the matching.
+    
+    bool doShowerKt = m_pythiaSignal->settings.flag("JetMatching:doShowerKt");
+    if (m_doMePsMatching && !doShowerKt) 
+       	njetNow = m_matching->nMEpartons().first;
+    //FIXME: "getProcessSubset()" method does not exist in < 8.219
+    //FIXME: simply un-comment the following two lines in >= 8.219
```
Comment (hegner) [inline Generation/src/components/PythiaInterface.cpp]: maybe we should just upgrade there...

File: Generation/src/components/PythiaInterface.cpp [hunk 2]
```diff
@@ -80,13 +137,88 @@ StatusCode PythiaInterface::execute() {
...
+  }
+
   // Reset the counter to count failed events in a row
   m_iAbort=0;
 
   // Print debug: Pythia event info
   if (msgLevel() <= MSG::DEBUG) {
 
     for (int i = 0; i < m_pythiaSignal->event.size(); ++i){
+      info () << "PythiaInterface Pythia8 abort : "<< m_iAbort << "/" << m_nAbort << endmsg;
```
Comment (hegner) [inline Generation/src/components/PythiaInterface.cpp]: protecting an info statement within an if debug statement (see line 218) doesn't look correct.

File: Generation/src/components/PythiaInterface.cpp [hunk 3]
```diff
@@ -54,6 +66,51 @@ StatusCode PythiaInterface::initialize() {
...
+  if (m_doMePsMatching) {
+    m_matching = std::unique_ptr<Pythia8::JetMatchingMadgraph>(new Pythia8::JetMatchingMadgraph());
+    if (!m_matching) {
+      return Error ( " Failed to initialise jet matching structures.");
+    }
+    m_pythiaSignal->setUserHooksPtr(m_matching.get());
+  }
+
+  // store additional branch containing matching/merging validation observables
+  if(m_doMePsMatching || m_doMePsMerging) declareOutput("mePsMatchingVars" , m_handleMePsMatchingVars, "mePsMatchingVars");
```
Comment (hegner) [inline Generation/src/components/PythiaInterface.cpp]: I'd prefer braces and a new line after the if statement - reducing the line length.

**Comment by jlingema** [comment]
@hegner: Do we want to maintain the LHE files here or only keep a simple example for testing in the repository?

**Comment by hegner** [comment]
@jlingema - if we want to do production sometime, we need to record them somewhere. May deserve a special place though in comparison to files for pure testing.

File: Generation/src/components/PythiaInterface.cpp (hunk 3 above)
Comment (selvaggi) [inline Generation/src/components/PythiaInterface.cpp]: addressed

File: Generation/src/components/PythiaInterface.cpp [hunk 4]
```diff
@@ -54,6 +66,51 @@ StatusCode PythiaInterface::initialize() {
...
+
+  // Currently, only one scheme at a time is allowed.
+  if (m_doMePsMerging && m_doMePsMatching) {
+    return Error ( "Jet matching and merging cannot be used simultaneously!" );
+  }
+  
+  // Allow to set the number of additional partons dynamically.
+  if (m_doMePsMerging) {
+    // Store merging scheme.
+    int scheme = ( m_pythiaSignal->settings.flag("Merging:doUMEPSTree")
```
Comment (selvaggi) [inline Generation/src/components/PythiaInterface.cpp]: addressed

File: Generation/src/components/PythiaInterface.cpp [hunk 5]
```diff
@@ -54,6 +66,51 @@ StatusCode PythiaInterface::initialize() {
...
+    m_pythiaSignal->setUserHooksPtr(m_setting.get());
+  }
+
+  // For jet matching, initialise the respective user hooks code.
+  if (m_doMePsMatching) {
+    m_matching = std::unique_ptr<Pythia8::JetMatchingMadgraph>(new Pythia8::JetMatchingMadgraph());
+    if (!m_matching) {
+      return Error ( " Failed to initialise jet matching structures.");
+    }
+    m_pythiaSignal->setUserHooksPtr(m_matching.get());
```
Comment (selvaggi) [inline Generation/src/components/PythiaInterface.cpp]: yes

**Comment by selvaggi** [comment]
we need this extra (small) LHE file for testing matching.

**Pull request merged**